core/src/main/java/org/elasticsearch/action/search/TransportMultiSearchAction.java

```diff
@@ -110,10 +117,11 @@ static int defaultMaxConcurrentSearches(int availableProcessors, ClusterState st
      * @param responseCounter incremented on each response
      * @param listener the listener attached to the multi-search request
      */
-    private void executeSearch(
+    void executeSearch(
             final Queue<SearchRequestSlot> requests,
             final AtomicArray<MultiSearchResponse.Item> responses,
             final AtomicInteger responseCounter,
+            long startTimeInNanos,
             final ActionListener<MultiSearchResponse> listener) {
```

Comment:
Let's change the signature of this method to accept an `ActionListener<Item[]>`, and then invoke this method by defining a new `listener` that wraps the `ActionListener<MultiSearchResponse>` and takes care of creating the `MutliSearchResponse` with the took time. This removes the need for threading `startTimeInNanos` all the way through execute search which I think simplifies things.